### PR TITLE
Don't die on NoClassDefFoundError

### PR DESCRIPTION
**Pull request by bratseth**

### CODE REVIEW

**Comment by yngveaasheim** [comment]
This is related to VESPA-11525.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 1]
```diff
@@ -52,27 +54,38 @@ public void deconstruct(Object component) {
...
+                    // Randomize restart over 10 minutes to avoid simultaneous cluster restarts
+                    Thread.sleep(random() * 1000 * 60 * 10);
+                }
+                catch (InterruptedException exception) {
+                    log.log(WARNING, "Randomized wait before dying disrupted. Dying now.");
+                }
                 com.yahoo.protect.Process.logAndDie("Error when deconstructing " + component, e);
-            } catch (Exception e) {
-                log.log(WARNING, "Exception thrown when deconstructing " + component, e);
-            } catch (Throwable t) {
```
Comment (baldersheim) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Should you not still catch Throwable ? Exception, NoClassDefFoundError and Error does not cover the whole world

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 2]
```diff
@@ -52,27 +54,38 @@ public void deconstruct(Object component) {
...
                 try {
-                    Thread.sleep((long) (new Random(System.nanoTime()).nextDouble() * 180 * 1000));
-                } catch (InterruptedException exception) { }
+                    // Randomize restart over 10 minutes to avoid simultaneous cluster restarts
+                    Thread.sleep(random() * 1000 * 60 * 10);
+                }
+                catch (InterruptedException exception) {
+                    log.log(WARNING, "Randomized wait before dying disrupted. Dying now.");
+                }
                 com.yahoo.protect.Process.logAndDie("Error when deconstructing " + component, e);
```
Comment (jobergum) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Add shutting down or something to the log message. "Shutting down due to error when deconstruction .. "

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 1 above)
Comment (bratseth) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Exception and Error does cover the whole world.

 .... unless we live in an unimaginably evil one. Ok, I guess that's worth hedging against.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 2 above)
Comment (bratseth) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Done.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 3]
```diff
@@ -52,27 +54,42 @@ public void deconstruct(Object component) {
...
+            }
+            catch (Exception | NoClassDefFoundError e) { // May get class not found due to it being already unloaded
                 log.log(WARNING, "Exception thrown when deconstructing " + component, e);
-            } catch (Throwable t) {
-                log.log(WARNING, "Unexpected Throwable thrown when deconstructing " + component, t);
+            }
+            catch (Error e) {
+                try {
+                    // Randomize restart over 10 minutes to avoid simultaneous cluster restarts
+                    Thread.sleep(random() * 1000 * 60 * 10);
```
Comment (gjoranv) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: I suggest to log both before and after sleeping.

**Review by gjoranv — CHANGES_REQUESTED** [review]

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 3 above)
Comment (bratseth) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Ok, done. In this module we do have LogLevel available so this allowed me to log this as FATAL as well.

Comment (gjoranv) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Have you remembered to push it? I can't see the change. Github is very slow right now though.

**Comment by bratseth** [comment]
Pushed now  :-)

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 4]
```diff
@@ -52,27 +55,44 @@ public void deconstruct(Object component) {
...
+            catch (Exception | NoClassDefFoundError e) { // May get class not found due to it being already unloaded
                 log.log(WARNING, "Exception thrown when deconstructing " + component, e);
-            } catch (Throwable t) {
-                log.log(WARNING, "Unexpected Throwable thrown when deconstructing " + component, t);
+            }
+            catch (Error e) {
+                try {
+                    // Randomize restart over 10 minutes to avoid simultaneous cluster restarts
+                    long randomSleepSeconds = random() * 60 * 10;
+                    log.log(LogLevel.FATAL, "Error when deconstructing " + component + " Will sleep for " +
```
Comment (gjoranv) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Nitpicking, but you forgot to put a '.' after the component.  `(". Will sleep...)`

**Review by gjoranv — APPROVED** [review]

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java [hunk 5]
```diff
@@ -52,27 +55,44 @@ public void deconstruct(Object component) {
...
                 component.deconstruct();
                 log.info("Finished deconstructing " + component);
-            } catch (Error e) {
-                try {
-                    Thread.sleep((long) (new Random(System.nanoTime()).nextDouble() * 180 * 1000));
-                } catch (InterruptedException exception) { }
-                com.yahoo.protect.Process.logAndDie("Error when deconstructing " + component, e);
-            } catch (Exception e) {
+            }
+            catch (Exception | NoClassDefFoundError e) { // May get class not found due to it being already unloaded
```
Comment (gjoranv) [inline container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: This is a resource leak, as the component might not have freed it's resources. However, including NoClassDef error doesn't make it that much worse. But it's something to look for if the container runs out of memory.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 4 above)
Comment (bratseth) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Fixed.

File: container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java (hunk 5 above)
Comment (bratseth) [reply container-disc/src/main/java/com/yahoo/container/jdisc/component/Deconstructor.java]: Agree on both points.

**Pull request merged**